File: backend/app/services/masthead_name_fit.py
"""Render-only fitting for legacy point-text masthead names.

New editor payloads already contain browser-measured geometry. This fallback
repairs pre-fitting Slate/Monument documents with ReportLab metrics without
mutating saved content or replacing the browser's typography decisions.
"""
from __future__ import annotations

from copy import deepcopy
import math

from app.services.pdf_generator import PDF_Generator


def _value(element, key, default=None):
    return element.get(key, default) if isinstance(element, dict) else getattr(element, key, default)


def _assign(element, key, value):
    if isinstance(element, dict):
        element[key] = value
    else:
        setattr(element, key, value)


def _number(value, default=0.0):
    try:
        result = float(value)
    except (TypeError, ValueError):
        return default
    return result if math.isfinite(result) else default


def fit_legacy_masthead_names(elements, *, page_width=595.0):
    """Return isolated fitted elements for managed names lacking fit metadata.

    A name must have both a matching identity descriptor and its contact band;
    ordinary freeform text and the templates' existing textarea names remain
    unchanged. Fit the largest hundredth-point font down to 14 pt, then wrap
    complete text and move downstream first-page content by the added height.
    Modern browser-fitted names keep their exact size, height and line records.
    The input element list and all nested descriptor dictionaries stay intact.
    """
    candidates = [element for element in elements if
                  _value(element, "category") == "text"
                  and _value(element, "mastheadRole") == "name"
                  and not _value(element, "deleted", False)
                  and not isinstance(_value(element, "nameFit"), dict)]
    if not candidates:
        return elements
    result = deepcopy(elements)
    measure = PDF_Generator.__new__(PDF_Generator)
    pagination_template = None
    for name in result:
        if (_value(name, "category") != "text"
                or _value(name, "mastheadRole") != "name"
                or _value(name, "deleted", False)
                or isinstance(_value(name, "nameFit"), dict)):
            continue
        band_id = _value(name, "mastheadBandId")
        identity_element = next((element for element in result if
                                 isinstance(_value(element, "mastheadIdentity"), dict)
                                 and _value(element, "mastheadIdentity").get("id") == band_id), None)
        if identity_element is None:
            continue
        identity = _value(identity_element, "mastheadIdentity")
        contact_id = identity.get("contactBandId")
        contact_element = next((element for element in result if
                                isinstance(_value(element, "contactBand"), dict)
                                and _value(element, "contactBand").get("id") == contact_id), None)
        if contact_element is None:
            continue
        parked_contact = _value(contact_element, "profilePhotoMainContactBand")
        contact = parked_contact if isinstance(parked_contact, dict) else _value(contact_element, "contactBand")
        anchor = contact.get("anchor") or {}
        left, top = _number(_value(name, "left")), _number(_value(name, "top"))
        base = max(0.01, _number(_value(name, "fontSize"), 24.0))
        page = _number(_value(name, "page"), 1)
        right = min(_number(page_width, 595.0) - 24.0,
                    _number(anchor.get("rightLimit"), _number(page_width, 595.0) - 24.0))
        for element in result:
            photo_left = _number(_value(element, "left"))
            photo_top = _number(_value(element, "top"))
            if (_value(element, "photoSlot")
                    and not _value(element, "photoSlotHidden", False)
                    and not _value(element, "deleted", False)
                    and _number(_value(element, "page"), 1) == page
                    and photo_left > left
                    and photo_top <= top + base
                    and photo_top + _number(_value(element, "height")) >= top - base):
                right = min(right, photo_left - 12.0)
        width = right - left
        authored_width = _number(_value(name, "width"))
        if authored_width > 0:
            width = min(width, authored_width)
        if width <= 1:
            continue

        content = str(_value(name, "content", "") or "")
        if _value(name, "textTransform") == "uppercase":
            content = content.upper()
        family = _value(name, "fontFamily") or "Inter"
        bold, italic = bool(_value(name, "bold")), bool(_value(name, "italic"))
        tracking = _number(_value(name, "letterSpacing"))
        prepared = measure._prepare_styled(content, _value(name, "runs"), bold, italic, False, None)
        font, _, _ = measure._resolve_font(family, bold, italic)

        def measured_width(size):
            if prepared is not None:
                clean, styles = prepared
                return measure._styled_run_width(clean, styles, family, size, tracking)
            return measure._line_width(content, font, size, tracking)

        minimum = min(base, 14.0)
        if measured_width(base) <= width - 1.0 and "\n" not in content:
            fitted = base
        else:
            low, high = math.ceil(minimum * 100), math.floor(base * 100)
            # Search integer hundredths to avoid rounding a measured fit back
            # above the available width at the final serialization boundary.
            while low < high:
                middle = (low + high + 1) // 2
                if measured_width(middle / 100.0) <= width - 1.0 and "\n" not in content:
                    low = middle
                else:
                    high = middle - 1
            fitted = low / 100.0
        line_height = round(fitted * 1.2, 2)
        if prepared is not None:
            clean, styles = prepared
            rows = len(measure._wrap_textarea_styled(
                clean, styles, family, fitted, tracking, width,
                (bold, italic, False, None), False,
            ))
        else:
            rows = len(measure._wrap_textarea(content, font, fitted, tracking, width))
        rows = max(1, rows)
        extra_height = round((rows - 1) * line_height, 2)
        for key, value in {
            "fontSize": fitted, "width": width, "height": round(rows * line_height, 2),
            "lineHeight": line_height,
            "nameFit": {"baseFontSize": base, "fittedFontSize": fitted,
                        "width": width, "extraHeight": extra_height},
        }.items():
            _assign(name, key, value)
        if extra_height == 0:
            continue

        # Only authored downstream flow moves: photo/sidebar/page chrome and
        # other pages retain their coordinates. Descriptor blueprints follow
        # the same displacement so later restore/reflow actions stay coherent.
        for element in result:
            if (element is not name and _number(_value(element, "page"), 1) == page
                    and not _value(element, "fixedToPage", False)
                    and not _value(element, "photoSlot")
                    and not _value(element, "deleted", False)
                    and _value(element, "flowRole") != "photo-contact-header"
                    and not (parked_contact and _value(element, "contactBandId") == contact_id
                             and _value(element, "contactChannel"))
                    and _value(element, "flowLane") != "sidebar"
                    and not str(_value(element, "flowRole", "")).startswith("sidebar")
                    and _number(_value(element, "top")) > top):
                _assign(element, "top", _number(_value(element, "top")) + extra_height)
        title = identity.get("title") or {}
        for spec in [title.get("spec"), *(title.get("decorations") or [])]:
            if isinstance(spec, dict) and "top" in spec:
                spec["top"] = _number(spec["top"]) + extra_height
        if "startY" in anchor:
            anchor["startY"] = _number(anchor["startY"]) + extra_height
        flow = contact.get("flow") or {}
        if "bodyTop" in flow:
            flow["bodyTop"] = _number(flow["bodyTop"]) + extra_height
        for element in result:
            blueprint = _value(element, "profilePhotoMainMastheadIdentity")
            if isinstance(blueprint, dict) and blueprint.get("id") == band_id:
                parked_title = blueprint.get("title") or {}
                for spec in [parked_title.get("spec"), *(parked_title.get("decorations") or [])]:
                    if isinstance(spec, dict) and "top" in spec:
                        spec["top"] = _number(spec["top"]) + extra_height
        pagination_template = "monument" if band_id == "monument-masthead" else "slate"
    if pagination_template is not None:
        _carry_overflow_records(result, template=pagination_template)
    return result


def _carry_overflow_records(elements, *, template):
    """Carry newly displaced main records below the existing footer boundary.

    The generators already split records larger than a complete content page.
    Keep each remaining page-scoped flowGroup atomic and attach section chrome
    to its first record. Reuse authored gaps, preserve every existing page's
    inset, and append overflow ahead of that page's existing records. Only new
    continuation pages receive cloned repeatable page chrome; portraits and
    first-page ornaments never repeat. This operates on the isolated render
    copy and does not rewrite the stored document or its declared page count.
    """
    from app.services.cv_generator_primitives import CONTENT_BOTTOM

    def is_main(element):
        role = str(_value(element, "flowRole", ""))
        return (not _value(element, "deleted", False)
                and not _value(element, "fixedToPage", False)
                and not _value(element, "photoSlot")
                and not _value(element, "contactChannel")
                and not _value(element, "contactBand")
                and not _value(element, "mastheadIdentity")
                and _value(element, "flowLane") != "sidebar"
                and not role.startswith("sidebar")
                and not role.startswith("masthead")
                and role != "photo-contact-header")

    def bottom(element):
        height = _number(_value(element, "height"))
        if height <= 0 and _value(element, "category") == "text":
            height = _number(_value(element, "fontSize")) * 0.6
        return _number(_value(element, "top")) + max(0, height)

    body = sorted((element for element in elements if is_main(element)),
                  key=lambda element: (_number(_value(element, "page"), 1),
                                       _number(_value(element, "top"))))
    if not any(bottom(element) > CONTENT_BOTTOM + 0.01 for element in body):
        return
    groups = {}
    for element in body:
        group = _value(element, "flowGroup")
        if group:
            groups.setdefault((_number(_value(element, "page"), 1), group), []).append(element)

    units, pending, consumed = [], [], set()
    for element in body:
        if id(element) in consumed:
            continue
        if _value(element, "flowRole") == "section-chrome":
            pending.append(element)
            consumed.add(id(element))
            continue
        key = (_number(_value(element, "page"), 1), _value(element, "flowGroup"))
        members = groups.get(key, [element])
        members = [member for member in members if id(member) not in consumed]
        members = pending + members
        pending = []
        consumed.update(id(member) for member in members)
        units.append(members)
    if pending:
        units.append(pending)

    # Un-grouped grid cells and decorative chip pairs can share a vertical
    # interval. Carry that complete row together instead of stacking cells
    # which were authored side by side.
    merged = []
    for members in units:
        previous = merged[-1] if merged else None
        if (previous
                and _number(_value(previous[0], "page"), 1) == _number(_value(members[0], "page"), 1)
                and min(_number(_value(member, "top")) for member in members)
                < max(bottom(member) for member in previous) - 0.01):
            previous.extend(members)
        else:
            merged.append(members)
    units = merged

    continuation_top = 72.0 if template == "monument" else 58.0
    current_page, cursor, previous_source_page, previous_source_bottom = 1, 0.0, None, 0.0
    original_pages = {_number(_value(element, "page"), 1) for element in elements}
    for members in units:
        source_page = int(_number(_value(members[0], "page"), 1))
        source_top = min(_number(_value(member, "top")) for member in members)
        source_bottom = max(bottom(member) for member in members)
        height = source_bottom - source_top
        gap = max(0.0, source_top - previous_source_bottom) if source_page == previous_source_page else 10.0
        if source_page > current_page:
            current_page, cursor = source_page, 0.0
        desired_top = max(source_top if source_page == current_page else continuation_top,
                          cursor + gap if cursor else source_top)
        if desired_top + height > CONTENT_BOTTOM + 0.01:
            current_page += 1
            desired_top = continuation_top
        for member in members:
            _assign(member, "top", round(_number(_value(member, "top")) + desired_top - source_top, 2))
            _assign(member, "page", current_page)
        cursor = desired_top + height
        previous_source_page, previous_source_bottom = source_page, source_bottom

    # Existing continuation chrome is the preferred blueprint; page one is a
    # fallback for a previously single-page document. Explicit repeat flags
    # preserve the template author's distinction between rails and masthead.
    chrome_source_page = min((page for page in original_pages if page > 1), default=1)
    chrome = [element for element in elements if
              _number(_value(element, "page"), 1) == chrome_source_page
              and _value(element, "fixedToPage", False)
              and not _value(element, "photoSlot")
              and not _value(element, "deleted", False)
              and _value(element, "repeatOnContinuation", True) is not False]
    added_chrome = []
    for page in range(1, current_page + 1):
        if page in original_pages:
            continue
        clones = deepcopy(chrome)
        for element in clones:
            _assign(element, "page", page)
            element_id = _value(element, "element_id")
            if element_id is not None:
                _assign(element, "element_id", f"{element_id}-name-fit-{page}")
            content = str(_value(element, "content", "") or "")
            if (_value(element, "category") == "text" and content.isdigit()
                    and _number(_value(element, "top")) > CONTENT_BOTTOM):
                _assign(element, "content", str(page).zfill(len(content)))
        added_chrome.extend(clones)
    # render_elements preserves authored drawing order. Backgrounds appended
    # after carried records would paint over their content on a new page.
    elements[0:0] = added_chrome

File: frontend/src/utils/mastheadNameFit.js
/**
 * Fit legacy point-text masthead names into their authored contact column.
 * Slate and Monument use this representation; the other templates already
 * own bounded multiline identity layouts. Geometry is persisted in A4 points,
 * never screen pixels, so zoom cannot change the fit or the exported document.
 */
import { styledSegments } from './textRuns.js';
import { measureTextareaHeight } from './textareaHeight.js';
import { isSidebarLaneElement, listDocumentSections, packDocumentSections, sectionElementIds } from './sectionStructure.js';
import { reconcileDocumentPages } from './structureOperation.js';

export const MASTHEAD_NAME_MIN_FONT_SIZE = 14;
const round = (value) => Math.round(value * 100) / 100;
const pageOf = (element) => Number(element.page) || 1;

/**
 * Apply an explicit properties-panel size as the name's new maximum, even if
 * the chosen value equals its current automatic fit. Typing and other patches
 * preserve the previous maximum. Returns a new element without mutating input.
 */
export function applyMastheadNameFontIntent(element, patch) {
  const fontSize = Number(patch.fontSize);
  return { ...element, ...patch,
    ...(element.nameFit && 'fontSize' in patch && Number.isFinite(fontSize) && fontSize > 0
      ? { nameFit: { ...element.nameFit, baseFontSize: fontSize, fittedFontSize: fontSize } } : {}),
  };
}

/**
 * Resolve the safe right edge from authored masthead geometry, including legacy
 * saved documents without fitting metadata. A smaller manually authored frame
 * remains authoritative. Photo bounds are considered only beside the name.
 */
function nameWidth(elements, name, contact, pageWidth) {
  let right = Math.min(pageWidth - 24, Number(contact.anchor.rightLimit));
  if (!Number.isFinite(right)) return null;
  for (const element of elements) {
    if (!element.photoSlot || element.photoSlotHidden || element.deleted
      || pageOf(element) !== pageOf(name) || Number(element.left) <= Number(name.left)) continue;
    const bottom = Number(element.top) + Number(element.height);
    if (Number(element.top) <= Number(name.top) + Number(name.fontSize)
      && bottom >= Number(name.top) - Number(name.fontSize)) {
      right = Math.min(right, Number(element.left) - 12);
    }
  }
  const authoredWidth = Number(name.width) > 0 && name.width !== name.nameFit?.width
    ? Number(name.width) : Number(name.nameFit?.maxWidth);
  if (authoredWidth > 0) {
    right = Math.min(right, Number(name.left) + authoredWidth);
  }
  return right > Number(name.left) ? round(right - Number(name.left)) : null;
}

/**
 * Compute a font size and line-box height without changing source text or runs.
 * The optional browser line resolver supplies exact shaped wraps; headless
 * callers use the same width/word fallback as other canvas text measurements.
 */
function fitName(name, width, measureTextWidth, measureLines) {
  const previous = name.nameFit;
  const maxWidth = Number(name.width) > 0 && name.width !== previous?.width
    ? Number(name.width) : previous?.maxWidth;
  const baseFontSize = Math.max(1, Number(previous && name.fontSize === previous.fittedFontSize
    ? previous.baseFontSize : name.fontSize) || 24);
  const floor = Math.min(baseFontSize, MASTHEAD_NAME_MIN_FONT_SIZE);
  const content = String(name.content ?? '');
  const transform = (text) => name.textTransform === 'uppercase' ? text.toUpperCase() : text;
  const segments = styledSegments(content, name.runs);
  const measure = (size) => segments.reduce((total, segment) => {
    const text = transform(segment.text);
    const style = { ...name, fontSize: size,
      bold: name.bold || segment.bold, italic: name.italic || segment.italic };
    // The shared canvas reader omits final tracking; CSS applies it after the
    // last character too. Include it per span, matching the PDF draw advances.
    return total + (measureTextWidth
      ? measureTextWidth(text, style) + (text ? Number(name.letterSpacing) || 0 : 0)
      : text.length * (size * 0.6 + (Number(name.letterSpacing) || 0)));
  }, 0);
  let size = baseFontSize;
  if (measure(size) > width - 1) {
    let low = floor;
    let high = baseFontSize;
    // Width is monotonic for a fixed face and tracking. A bounded binary search
    // keeps typing cheap even for large display sizes; round down to stay inside.
    for (let step = 0; step < 16; step += 1) {
      const middle = (low + high) / 2;
      if (measure(middle) <= width - 1) low = middle;
      else high = middle;
    }
    size = Math.max(floor, Math.floor(low * 100) / 100);
  }
  const lineHeight = round(size * 1.2);
  const probe = { ...name, category: 'textarea', width, fontSize: size, lineHeight };
  let rows = 1;
  // Fitting does not remove authored line breaks. Even at the minimum an
  // unbroken surname can wrap; no ellipsis or substring is stored or exported.
  if (measure(size) > width - 1 || /[\r\n]/.test(content)) {
    const lines = measureLines?.(probe);
    rows = lines?.length || Math.max(1, Math.round((measureTextareaHeight(
      transform(content), width, size, lineHeight,
      { measureTextWidth, textStyle: probe },
    ) - 6) / lineHeight));
  }
  const extraHeight = round((rows - 1) * lineHeight);
  return { fontSize: size, width, height: round(rows * lineHeight), lineHeight,
    nameFit: { baseFontSize, fittedFontSize: size, width, extraHeight,
      ...(maxWidth > 0 ? { maxWidth } : {}),
    } };
}

function shiftIdentityDescriptor(identity, delta) {
    const title = identity.title;
    return { ...identity,
      title: { ...title,
        ...(title?.spec ? { spec: { ...title.spec, top: round(Number(title.spec.top) + delta) } } : {}),
        ...(title?.decorations ? { decorations: title.decorations.map((decoration) => ({
          ...decoration, top: round(Number(decoration.top) + delta),
        })) } : {}),
      },
    };
}

/** Shift the managed main lane while keeping photo, sidebar and page chrome fixed. */
function shiftBelowName(element, name, identity, contactAnchor, delta) {
  if (!delta || element === name || pageOf(element) !== pageOf(name)) return element;
  if (element === identity) {
    return { ...element, mastheadIdentity: shiftIdentityDescriptor(element.mastheadIdentity, delta),
      ...(element.profilePhotoMainMastheadIdentity ? {
        profilePhotoMainMastheadIdentity: shiftIdentityDescriptor(element.profilePhotoMainMastheadIdentity, delta),
      } : {}),
    };
  }
  if (element === contactAnchor) {
    // Slate parks the main descriptor while contacts occupy the photo-less
    // rail. Move its restoration target, leaving the visible sidebar untouched.
    const key = element.profilePhotoMainContactBand ? 'profilePhotoMainContactBand' : 'contactBand';
    const band = element[key];
    return { ...element, [key]: { ...band,
      anchor: { ...band.anchor, startY: round(Number(band.anchor.startY) + delta) },
      ...(band.flow ? { flow: { ...band.flow, bodyTop: round(Number(band.flow.bodyTop) + delta) } } : {}),
    } };
  }
  if (element.fixedToPage || element.photoSlot || element.flowLane === 'sidebar'
    || String(element.flowRole || '').startsWith('sidebar')
    || element.flowRole === 'photo-contact-header'
    || (contactAnchor.profilePhotoMainContactBand && element.contactChannel
      && element.contactBandId === contactAnchor.contactBand.id)
    || Number(element.top) <= Number(name.top)) return element;
  return { ...element, top: round(Number(element.top) + delta) };
}

/**
 * Normalize complete document elements before a canvas state transaction.
 *
 * Only semantic point-text names with a coupled contact column opt in. The
 * stored base font is the user's size intent; the effective font can grow back
 * after shortening. Explicit font edits replace that base. Saved extra height
 * makes wrapping/shortening reversible without repeatedly shifting the page.
 * Neither source strings, inline formatting nor unrelated templates are changed.
 *
 * @param {object[]} elements - Current authored graph, never mutated.
 * @param {object} options - Width/line readers and physical page width.
 * @returns {object[]} Fitted graph, preserving the array on an unchanged layout.
 */
export function fitMastheadNames(elements, {
  measureTextWidth = null, measureLines = null, pageWidth = 595,
  pageHeight = 842, spacing, createId = () => globalThis.crypto.randomUUID(),
} = {}) {
  if (!Array.isArray(elements)) return elements;
  let result = elements;
  for (const candidate of elements) {
    if (candidate.category !== 'text' || candidate.mastheadRole !== 'name' || candidate.deleted) continue;
    const name = result.find((element) => element === candidate) || candidate;
    const identity = result.find((element) => element.mastheadIdentity
      && element.mastheadBandId === name.mastheadBandId);
    const contactAnchor = result.find((element) => element.contactBand
      && element.contactBand.id === identity?.mastheadIdentity.contactBandId);
    if (!identity || !contactAnchor) continue;
    const width = nameWidth(result, name,
      contactAnchor.profilePhotoMainContactBand || contactAnchor.contactBand, pageWidth);
    if (!width) continue;
    const fitted = fitName(name, width, measureTextWidth, measureLines);
    if (name.nameFit?.repaginate) fitted.nameFit.repaginate = true;
    if (Number.isFinite(name.nameFit?.flowStart)) fitted.nameFit.flowStart = name.nameFit.flowStart;
    const delta = round(fitted.nameFit.extraHeight - (Number(name.nameFit?.extraHeight) || 0));
    if (['fontSize', 'width', 'height', 'lineHeight'].every((key) => name[key] === fitted[key])
      && JSON.stringify(name.nameFit) === JSON.stringify(fitted.nameFit)) continue;
    const before = result;
    result = result.map((element) => element === name ? { ...name, ...fitted }
      : shiftBelowName(element, name, identity, contactAnchor, delta));
    const overflow = delta > 0 && result.some((element) => !element.fixedToPage
      && !element.photoSlot && !isSidebarLaneElement(element)
      && element.flowRole !== 'masthead' && element.flowRole !== 'masthead-anchor'
      && Number(element.top) + (element.category === 'text'
        ? Number(element.fontSize) * 0.7 : Number(element.height) || 0) > pageHeight - 72);
    if (delta && (overflow || name.nameFit?.repaginate)) {
      // A taller identity may push an otherwise full page into its footer.
      // Use the existing record-aware packer and pre-shift membership so no
      // role is split or accidentally attached to a later section. Subsequent
      // shortening also reclaims continuation space; ordinary fits keep every
      // body coordinate unchanged. Reconciliation adds/removes page chrome.
      const sections = listDocumentSections(before, pageHeight);
      const firstIds = sections.length ? sectionElementIds(before, sections[0].headingId, pageHeight) : new Set();
      const firstChrome = before.filter((element) => firstIds.has(element.element_id)
        && element.flowRole === 'section-chrome');
      const originalStart = name.nameFit?.flowStart ?? Math.min(...firstChrome.map((element) =>
        (pageOf(element) - 1) * pageHeight + Number(element.top)));
      const flowStart = originalStart + delta;
      // Preserve the authored first-section clearance. The packer's generic
      // legacy-gap repair would otherwise mistake a wrapped masthead for a
      // stale gap and move Slate's first heading upwards after shortening.
      const anchored = Number.isFinite(flowStart) ? result.map((element) => element.element_id === identity.element_id
        ? { ...element, mainFlowStart: flowStart } : element) : result;
      result = packDocumentSections(anchored, sections.map((section) => section.headingId), pageHeight,
      { spacing, membershipReference: before });
      result = reconcileDocumentPages(result, createId, { collapseEmpty: true }).elements;
      result = result.map((element) => {
        if (element.element_id === name.element_id) return { ...element,
          nameFit: { ...element.nameFit, repaginate: true,
            ...(Number.isFinite(flowStart) ? { flowStart } : {}),
          } };
        if (element.element_id === identity.element_id && Number.isFinite(flowStart)) {
          const { mainFlowStart: _ignored, ...rest } = element;
          return 'mainFlowStart' in identity ? { ...rest, mainFlowStart: identity.mainFlowStart } : rest;
        }
        return element;
      });
    }
  }
  return result;
}
